### DISCUSSION TITLE
`Shared` subscription issue

### DISCUSSION THREAD

**Discussion by ddanilyuk** [pointfreeco/swift-composable-architecture#3467] [General]
### Description

Found some weird behavior with AppStorageKey. The shared `publisher` is triggered after updating some unrelated UserDefaults values.
Here is the XCTest, which is failing but should not.
``` swift
func testSharedSubscription() throws {
    @Dependency(\.defaultAppStorage) var defaults
    var cancellable: (any Cancellable)?
    @Shared(.appStorage("testInt")) var testInt: Int = 0
    var isUpdated = false
    cancellable = $testInt.publisher.sink { _ in
        isUpdated = true
    }
    defaults.set(42, forKey: "randomDefaultsKey")
    XCTAssert(!isUpdated)
    _ = cancellable
}
```

### Checklist

- [X] I have determined whether this bug is also reproducible in a vanilla SwiftUI project.
- [X] If possible, I've reproduced the issue using the `main` branch of this package.
- [X] This issue hasn't been addressed in an [existing GitHub issue](https://github.com/pointfreeco/swift-composable-architecture/issues) or [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions).

### Reproducing project

https://github.com/ddanilyuk/SharedSubscriptionIssue/blob/main/TestSharedSubscriptionTests/TestSharedSubscriptionTests.swift

### The Composable Architecture version information

1.15.2

### Xcode version information

16.0

**Comment by stephencelis** [comment]
@ddanilyuk This is the expected behavior as of today, and you'll see similar behavior in SwiftUI's `@AppStorage`. In order to subscribe to updates we use Notification Center because KVO does not work with user defaults keys that contain periods (_e.g._ `"my.key"`), and Notification Center does not allow for granular subscription per key. As such we just always trigger an update when user defaults update, just like `@AppStorage` in SwiftUI.

We are open to improvements here, but we don't consider it a bug, so I'm going to convert this to a discussion.
